public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx

```diff
@@ -0,0 +1,254 @@
+import React, { FC, useCallback, useEffect, useMemo, useReducer, useRef, useState } from 'react';
+import { useFormContext } from 'react-hook-form';
+
+import { LoadingState, PanelData } from '@grafana/data';
+import { selectors } from '@grafana/e2e-selectors';
+import { config } from '@grafana/runtime';
+import { Alert, Button, Field, InputControl, Stack, Tooltip } from '@grafana/ui';
+import { isExpressionQuery } from 'app/features/expressions/guards';
+import { AlertQuery } from 'app/types/unified-alerting-dto';
+
+import { AlertingQueryRunner } from '../../../state/AlertingQueryRunner';
+import { RuleFormType, RuleFormValues } from '../../../types/rule-form';
+import { getDefaultOrFirstCompatibleDataSource } from '../../../utils/datasource';
+import { ExpressionEditor } from '../ExpressionEditor';
+import { ExpressionsEditor } from '../ExpressionsEditor';
+import { QueryEditor } from '../QueryEditor';
+import { RuleEditorSection } from '../RuleEditorSection';
+import { refIdExists } from '../util';
+
+import { AlertType } from './AlertType';
+import {
+  duplicateQuery,
+  addNewDataQuery,
+  addNewExpression,
+  queriesAndExpressionsReducer,
+  removeExpression,
+  rewireExpressions,
+  setDataQueries,
+  updateExpression,
+  updateExpressionRefId,
+  updateExpressionType,
+} from './reducer';
+
+interface Props {
+  editingExistingRule: boolean;
+}
+
+export const QueryAndExpressionsStep: FC<Props> = ({ editingExistingRule }) => {
+  const runner = useRef(new AlertingQueryRunner());
+  const {
+    setValue,
+    getValues,
+    watch,
+    formState: { errors },
+    control,
+  } = useFormContext<RuleFormValues>();
+  const [panelData, setPanelData] = useState<Record<string, PanelData>>({});
+
+  const initialState = {
+    queries: getValues('queries'),
+    panelData: {},
+  };
+  const [{ queries }, dispatch] = useReducer(queriesAndExpressionsReducer, initialState);
+
+  const type = watch('type');
```

Comment:
You can try `[type, condition, dataSourceName = watch(['type', 'condition', 'dataSourceName]` to save a few lines